lgc/patch/PatchInOutImportExport.cpp

```diff
@@ -5667,6 +5532,75 @@ Value *PatchInOutImportExport::getDeviceIndex(Instruction *insertPos) {
   }
 }
 
+// =====================================================================================================================
+// Records export info of vertex attributes
+//
+// @param location : Vertex attribute location
+// @param attribValues : Values of this vertex attribute to export
+void PatchInOutImportExport::recordVertexAttribExport(unsigned location, ArrayRef<Value *> attribValues) {
+  assert(m_shaderStage == ShaderStageVertex || m_shaderStage == ShaderStageTessEval ||
+         m_shaderStage == ShaderStageCopyShader); // Valid shader stages
+  assert(attribValues.size() == 4);               // Must have 4 elements, corresponds to <4 x float>
+
+  auto undef = UndefValue::get(Type::getFloatTy(*m_context));
+
+  // Vertex attribute not existing, insert a new one and initialize it
+  if (m_attribExports.find(location) == m_attribExports.end()) {
+    for (unsigned i = 0; i < 4; ++i)
+      m_attribExports[location][i] = undef;
+  }
+
+  for (unsigned i = 0; i < 4; ++i) {
+    assert(attribValues[i]);
+    if (!isa<UndefValue>(attribValues[i])) {
+      assert(isa<UndefValue>(m_attribExports[location][i])); // Must be invalid, couldn't have overlapping
+      m_attribExports[location][i] = attribValues[i];        // Update values that are valid (not undef)
+    }
+  }
+
+  auto &inOutUsage = m_pipelineState->getShaderResourceUsage(m_shaderStage)->inOutUsage;
+  inOutUsage.expCount = std::max(inOutUsage.expCount, location + 1); // Update export count
+}
+
+// =====================================================================================================================
+// Exports vertex attributes that were recorded previously
+//
+// @param insertPos : Where to insert instructions.
+void PatchInOutImportExport::exportVertexAttribs(Instruction *insertPos) {
+  assert(m_shaderStage == ShaderStageVertex || m_shaderStage == ShaderStageTessEval ||
+         m_shaderStage == ShaderStageCopyShader); // Valid shader stages
+  if (m_attribExports.empty()) {
+    assert(m_pipelineState->getShaderResourceUsage(m_shaderStage)->inOutUsage.expCount == 0);
+    return;
+  }
+
+  IRBuilder<> builder(*m_context);
+  builder.SetInsertPoint(insertPos);
+
+  for (auto &attribExport : m_attribExports) {
+    if (m_gfxIp.major <= 10) {
```

Comment:
Replace the if/else with `assert(m_gfxIp.major <= 10 && "Not implemented for this GFX IP")`?

Reply:
We use such if/else structure for us to easily support new GPU internally. And this could match GitHub public codes consistently when doing porting.

Reply:
Ack. Makes sense.